### PR TITLE
fix: ignore failed outline API call when learner has no access

### PR DESCRIPTION
**Pull request by pwnage101**
If the learner doesn't even have access to the course (e.g. because the course starts in the future), don't worry about a 404 fetching the course outline since we're not planning to use it anyway.

ENT-8078

matrix
===

* [current behavior] if the learner has access to the course and tab data is available: stay on course home page.
* [current behavior] if the learner has access to the course but the tab data is not available, display error page.
* [current behavior] if the learner does not have access to the course, but the tab data is available, dispatch to enterprise redirect logic.
* [NEW BEHAVIOR] if the learner does not have access to the course, but the tab data is not available, dispatch to enterprise redirect logic (WAS: display error page).

A combination of factors contribute to this specific scenario with new behavior, but the vast majority of hits on this page won't trigger this scenario.  It just so happens that it's likely to happen while testing on stage, which gets in the way of testing other features which rely on dispatching to fetchTabDenied.

### CODE REVIEW

File: src/course-home/data/thunks.js [hunk 1]
```diff
@@ -38,28 +38,39 @@ export function fetchTab(courseId, tab, getTabData, targetUserId) {
...
+        throw courseHomeCourseMetadataResult.reason;
+      } else if (!courseHomeCourseMetadataResult.value.courseAccess.hasAccess) {
         dispatch(fetchTabDenied({ courseId }));
-      } else if (tabDataResult || !getTabData) {
+      } else if (tabDataResult && tabDataResult.status === 'rejected') {
+        throw tabDataResult.reason;
+      } else {
         dispatch(fetchTabSuccess({
           courseId,
           targetUserId,
```
Comment (pwnage101) [inline src/course-home/data/thunks.js]: Notable changes:
* There's now a default case; one of [fetchTabDenied, fetchTabSuccess, fetchTabFailure] is always dispatched.
* A non-accessible course is now handled using fetchTabDenied regardless of the tab data query results.

File: src/course-home/data/thunks.js [hunk 2]
```diff
@@ -38,28 +38,39 @@ export function fetchTab(courseId, tab, getTabData, targetUserId) {
...
-      // Disable the access-denied path for now - it caused a regression
-      if (!courseHomeCourseMetadata.courseAccess.hasAccess) {
+      if (courseHomeCourseMetadataResult.status === 'rejected') {
+        throw courseHomeCourseMetadataResult.reason;
+      } else if (!courseHomeCourseMetadataResult.value.courseAccess.hasAccess) {
         dispatch(fetchTabDenied({ courseId }));
-      } else if (tabDataResult || !getTabData) {
+      } else if (tabDataResult && tabDataResult.status === 'rejected') {
+        throw tabDataResult.reason;
+      } else {
```
Comment (adamstankiewicz) [inline src/course-home/data/thunks.js]: nit: I wonder if it'd make sense to co-locate the conditions that result in throwing an exception?

```jsx
if (courseHomeCourseMetadataResult.status === 'rejected') {
  throw courseHomeCourseMetadataResult.reason;
} else if (tabDataResult?.status === 'rejected') {
  throw tabDataResult.reason;
} else if (!courseHomeCourseMetadataResult.value.courseAccess.hasAccess) {
 dispatch(fetchTabDenied({ courseId }));
} else {
  dispatch(fetchTabSuccess({
    courseId,
    targetUserId,
  }));
}
```

File: src/course-home/data/thunks.js [hunk 3]
```diff
@@ -38,28 +38,39 @@ export function fetchTab(courseId, tab, getTabData, targetUserId) {
...
+      if (courseHomeCourseMetadataResult.status === 'fulfilled') {
+        dispatch(addModel({
+          modelType: 'courseHomeMeta',
+          model: {
+            id: courseId,
+            ...courseHomeCourseMetadataResult.value,
+          },
+        }));
+      }
+      if (tabDataResult && tabDataResult.status === 'fulfilled') {
```
Comment (adamstankiewicz) [inline src/course-home/data/thunks.js]: nit:

```suggestion
      if (tabDataResult?.status === 'fulfilled') {
```

File: src/course-home/data/thunks.js [hunk 4]
```diff
@@ -38,28 +38,39 @@ export function fetchTab(courseId, tab, getTabData, targetUserId) {
...
         }));
       }
-      // Disable the access-denied path for now - it caused a regression
-      if (!courseHomeCourseMetadata.courseAccess.hasAccess) {
+      if (courseHomeCourseMetadataResult.status === 'rejected') {
+        throw courseHomeCourseMetadataResult.reason;
+      } else if (!courseHomeCourseMetadataResult.value.courseAccess.hasAccess) {
         dispatch(fetchTabDenied({ courseId }));
-      } else if (tabDataResult || !getTabData) {
+      } else if (tabDataResult && tabDataResult.status === 'rejected') {
```
Comment (adamstankiewicz) [inline src/course-home/data/thunks.js]: nit:

```suggestion
      } else if (tabDataResult?.status === 'rejected') {
```

**Review by adamstankiewicz — APPROVED** [review]
LGTM (with a couple nits), deferring to any owning squad's review feedback.

File: src/course-home/data/thunks.js (hunk 2 above)
Comment (pwnage101) [reply src/course-home/data/thunks.js]: Hm, I'm pretty sure I want to dispatch fetchTabDenied as soon as possible, which is why the conditional is ordered like that.  I added a comment to clarify this.

**Review by nsprenkle — APPROVED** [review]

**Pull request merged**